---
title: "CancerAbstract Vignette"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{CancerAbstract Vignette}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.width = 7, 
  fig.height = 7
)
```

```{r setup}
library(CancerAbstract)
library(ggplot2)
```

# Part I: Kriging Interpolation

Load the spatial transcriptomics data from package: 

- **ST_expr_data**, representing the spot-by-gene expression metrix normalized by Seurat::SCTransform. 

```{r}
ST_expr_data[1:5, 1:5]
```

- **spot_coord**, representing the physical location (coordinates) of every spots. It should have the same row names as the expression matrix. Here in the example, we have two columns 'imagerow' and 'imagecol' as X and Y coordinates in $\mu m$. 

```{r}
head(spot_coord)
```


### Apply Kriging method on ST data

As the first step of CancerAbstract, a Kriging interpolation is performed for a given gene. 

```{r, eval=FALSE}
Kriging_res <- SpatialKriging(spatial.data = ST_expr_data,  # spatial expression matrix
                              spatial.coord = spot_coord,   # coordinates
                              gene.id = "IFI6",             # gene id, a colname in spatial.data
                              predict.grid.size = 20,       # grid size for interpolation
                              Kriging.cutoff = 600,         # auotocorrelation effect range
                              Kriging.width = 30,           # bin widths in Kriging sampling
                              plot.dir = "./plots/Kriging/",# path to save figures
                              save.plot = F                 # plot or not; default is T
                              )
```

*SpatialKriging* contains 4 sub-functions that can be run separately as below:

- integrateCoordinate: based on sp::coordinates(), integrate the ST expression matrix (STEM) and coordiantes together into one object

- fitKriging: fit Kriging model on integrated data; should indicate specific gene id

- createGridDataframe: create the new grid based on the original spatial coordinates; so far only support creating a rectangle space with square tiles

- Kriging_predict: perform prediction / interpolation on newly created grid

```{r}
integrate_ST_data <- integrateCoordinate(ST_expr_data, spot_coord)
Kriging_model <- fitKriging(integrate_ST_data, "IFI6", save.plot = F)
finer_grid <- createGridDataframe(integrate_ST_data)
Kriging_predict <- predictKriging(integrate_ST_data,
                                  finer_grid,
                                  "IFI6",
                                  Kriging_model[[2]],
                                  save.plot = F)
Kriging_res <- data.frame(X = Kriging_predict@coords[,1],
                          Y = Kriging_predict@coords[,2],
                          var1.pred = Kriging_predict$var1.pred)
```

Visualization of the interpolation results

```{r}
ggplot(data = Kriging_res) + 
    geom_tile(aes(x = X, 
                  y = Y, 
                  fill = var1.pred)) + 
    scale_fill_gradient2(low = "darkblue", 
                         high = "yellow", 
                         mid = "purple", 
                         midpoint = max(Kriging_res$var1.pred, na.rm = T) / 2) +
    theme_bw() + 
    labs(x = "X / um", 
         y = "Y / um", 
         title = "IFI6 Kriging Interplocation", 
         fill = "Predicted Expression") +
    theme(aspect.ratio = 1)
```



